fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MetaDataProtoEditor.java

```diff
@@ -238,4 +256,79 @@ private static void setDeprecated(DescriptorProtos.FieldDescriptorProto.Builder
         return messageType.getFieldBuilderList().stream().filter(m -> m.getName().equals(fieldName)).findAny().orElse(null);
     }
 
+    /**
+     * Add a default union to the given records descriptor if missing.
+     *
+     * <p>
+     * This method is a no-op if the union is present. Otherwise, the method will add a union to the records
+     * descriptor. The union descriptor will be filled in with all of the record types defined in the file except
+     * {@code NESTED} record types.
+     * </p>
+     *
+     * @param fileDescriptor the records descriptor of the record meta-data
+     * @return the resulting records descriptor
+     */
+    @Nonnull
+    public static Descriptors.FileDescriptor addDefaultUnionIfMissing(Descriptors.FileDescriptor fileDescriptor) {
+        if (MetaDataProtoEditor.hasUnion(fileDescriptor)) {
+            return fileDescriptor;
+        }
+        DescriptorProtos.FileDescriptorProto fileDescriptorProto = fileDescriptor.toProto();
+        DescriptorProtos.FileDescriptorProto.Builder fileBuilder = fileDescriptorProto.toBuilder();
+        fileBuilder.addMessageType(MetaDataProtoEditor.createDefaultUnion(fileDescriptor));
+        try {
+            return Descriptors.FileDescriptor.buildFrom(fileBuilder.build(), fileDescriptor.getDependencies().toArray(new Descriptors.FileDescriptor[0]));
+        } catch (Descriptors.DescriptorValidationException e) {
+            throw new MetaDataException("Failed to add a default union", e);
+        }
+    }
+
+    @Nonnull
+    private static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor) {
+        DescriptorProtos.DescriptorProto.Builder unionMessageType = DescriptorProtos.DescriptorProto.newBuilder();
+        unionMessageType.setName(RecordMetaDataBuilder.DEFAULT_UNION_NAME);
+        for (Descriptors.Descriptor messageType : fileDescriptor.getMessageTypes()) {
+            RecordMetaDataOptionsProto.RecordTypeOptions.Usage messageTypeUsage = getMessageTypeUsage(messageType.toProto());
+            if (messageTypeUsage != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.NESTED) {
+                addFieldToUnion(unionMessageType, messageType.getName(), messageType.getFullName());
+            }
+        }
+        return unionMessageType;
+    }
+
+    /**
+     * Creates a default union descriptor for the given file descriptor and a base union descriptor. It adds all of the
+     * non-{@code NESTED} message types that exist in the base union to the synthetic union.
+     * @param fileDescriptor the file descriptor to create a union for
+     * @param baseUnionDescriptor the base union descriptor
+     * @return the builder for the union
+     */
+    @Nonnull
+    @API(API.Status.INTERNAL)
+    public static DescriptorProtos.DescriptorProto.Builder createDefaultUnion(@Nonnull Descriptors.FileDescriptor fileDescriptor, @Nonnull Descriptors.Descriptor baseUnionDescriptor) {
```

Comment:
Maybe `createSyntheticUnion` instead of `createDefaultUnion`? Or something to indicate that it's creating the union based on the templated union descriptor rather than from *tabula rasa*.

Reply:
sure